src/document/doc-utils.js

```diff
@@ -180,42 +180,132 @@ function removeLines(doc) {
   return mapDoc(doc, removeLinesFn);
 }
 
-function getInnerParts(doc) {
-  let { parts } = doc;
-  let lastPart;
-  // Avoid a falsy element like ""
-  for (let i = doc.parts.length; i > 0 && !lastPart; i--) {
-    lastPart = parts[i - 1];
+const isHardline = (doc, nextDoc) =>
+  doc &&
+  doc.type === "line" &&
+  doc.hard &&
+  nextDoc &&
+  nextDoc.type === "break-parent";
+function stripDocTrailingHardlineFromDoc(doc) {
+  if (!doc) {
+    return doc;
   }
-  if (lastPart.type === "group") {
-    parts = lastPart.contents.parts;
+
+  switch (doc.type) {
+    case "concat":
+    case "fill": {
+      const { parts } = doc;
+
+      while (parts.length > 1 && isHardline(...parts.slice(-2))) {
+        parts.length -= 2;
+      }
+
+      if (parts.length > 0) {
+        const lastPart = stripDocTrailingHardlineFromDoc(
+          parts[parts.length - 1]
+        );
+        parts[parts.length - 1] = lastPart;
+      }
+      break;
+    }
+    case "align":
+    case "indent":
+    case "group":
+    case "line-suffix":
+      doc.contents = stripDocTrailingHardlineFromDoc(doc.contents);
+      break;
+    case "if-break":
+      doc.breakContents = stripDocTrailingHardlineFromDoc(doc.breakContents);
+      doc.flatContents = stripDocTrailingHardlineFromDoc(doc.flatContents);
+      break;
   }
-  return parts;
+
+  return doc;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
+function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
-  if (doc.type === "concat" && doc.parts.length !== 0) {
-    const parts = withInnerParts ? getInnerParts(doc) : doc.parts;
-    const lastPart = parts[parts.length - 1];
-    if (lastPart.type === "concat") {
+  return stripDocTrailingHardlineFromDoc(cleanDoc(doc));
+}
+
+const isConcat = (doc) => doc && doc.type === "concat";
+function cleanDocFn(doc) {
+  switch (doc.type) {
+    case "fill":
+      if (doc.parts.length === 0 || doc.parts.every((part) => part === "")) {
+        return "";
+      }
+      break;
+    case "group":
+      // Remove nested only group
       if (
-        lastPart.parts.length === 2 &&
-        lastPart.parts[0].hard &&
-        lastPart.parts[1].type === "break-parent"
+        doc.contents.type === "group" &&
+        doc.contents.id === doc.id &&
+        doc.contents.break === doc.break &&
+        doc.contents.expandedStates === doc.expandedStates
```

Comment:
I don't understand this part. I thought you'd just need to replace `if (!doc.contents)` with `if (!doc.contents && !doc.expandedStates)`.

Reply:
For hardline searching it's not needed, but I try to make it safe  to reduce nesting groups, for later use.

You can format `.codecov.yml`, I'm shocked, JSON.stringify the doc  with 2 spaces, it's 20mb...

Reply:
Format this 128b `.codecov.yml` file,

Stringify the doc with

```js
	JSON.stringify(
      doc,
      (key, value) => (key === "expandedStates" ? undefined : value),
      2
    )
```

Original: 1803KB, with this code 266KB, without this 392KB

Reply:
It's better to measure and compare JSON without spaces.

But back to my question. Yes, let's keep hardline searching out of the equation and think of this as of a generic normalizing function. Is this condition ever true: `doc.contents.id === doc.id`? Aren't ids supposed to be unique? What's wrong with the replacement I mentioned in https://github.com/prettier/prettier/pull/9736#discussion_r531718688?

Reply:
group.id most time just "undefined", if there are not equal, it can't be removed, it will effect ifBreak, only if it's "undefined", removing the inner one will not effect the ifBreak, because they both decided by the parent group.

I didn't learn too much about docs, but I think this is how they works?

Reply:
> It's better to measure and compare JSON without spaces.

Original: 60KB, remove group 26KB, without removing group 34KB

Reply:
Indeed, I didn't think about the `undefined === undefined` case. `!doc.id && !doc.contents.id` looks like a cleaner way to express this logic. Anyway, it's also definitely safe to remove a group if it has empty `contents` (`""` or `concat([])`), no `id`, and no `expandedStates`.

Reply:
> !doc.id && !doc.contents.id

 Won't work if they have same symbol id, they can be removed too. :)

I tried to remove empty `contents` group, but test failed, maybe because I didn't check `id` and `expandedStates`

Reply:
"they have same symbol id" - Is this really possible? Sounds like a bug in the printer.

Yes, you have to check all the properties like I described above (note that I edited that comment).

Reply:
Don't know why yaml printer use so many groups.

Reply:
> "they have same symbol id" - Is this really possible? Sounds like a bug in the printer.


I think yes, but maybe not nested, in js printer there is getXXXGroupId function, it's cached, I'm in bed now, you can find it.